File: tools/spellFolders.py
import os
import json
import utils
import buildReferenceFolder

targetPack = "spells"
packPath = os.path.join(".\src", "packs", targetPack)
packList = os.listdir(packPath)

class Folder:
    def __init__(self, name: str, color: str | None = None, parentFolder: str | None = None) -> None:
        self.name = name
        self.color = color
        self.flags = {}
        self.sort = 0
        self.sorting = 'a'
        self._type = 'Item'
        self.folder = parentFolder
        self._id = utils.randomID()
        self._key = '!folders!' + self._id
        
    def __repr__(self) -> str:
        return repr(vars(self))

class encoder(json.JSONEncoder):
    def default(self, o):
        d: dict = o.__dict__
        if "_type" in d:
            d["type"] = d["_type"]
            d.pop("_type")
        return d

def createFolders():
    outerFolders = ["Spells", "Rare Spells"]
    innerFolders = ["Cantrip", "1st level", "2nd level", "3rd level", "4th level", "5th level", "6th level", "7th level", "8th level", "9th level"]
    for o in outerFolders:
        outer = Folder(o)
        print(outer)
        filename = outer.name.replace(' ', '_') + '_' + outer._id + ".json"
        with open(os.path.join(packPath,filename), "w") as writeFile:
            writeFile.write(json.dumps(outer, indent=2, cls=encoder))
        for i in innerFolders:
            inner = Folder(i, parentFolder=outer._id)
            print(inner)
            filename = inner.name.replace(' ', '_') + '_' + inner._id + ".json"
            with open(os.path.join(packPath,filename), "w") as writeFile:
                writeFile.write(json.dumps(inner, indent=2, cls=encoder))
    buildReferenceFolder.build(targetPack)

if not os.path.exists(os.path.join("tools", "referenceFolders", "spells.json")):
    createFolders()

spells: list[str] = list('0123456789')
rareSpells: list[str] = list('0123456789')

folders = utils.prepFolders(targetPack)



for f in folders:
    if folders[f]["parent"] != None:
        parentName: str = utils.folderName(folders, f, 1)
        currentName: str = folders[f]["name"]
        level: int = int(currentName[0]) if currentName != 'Cantrip' else 0
        if parentName == 'Spells':
            spells[level] = f
        else:
            rareSpells[level] = f

for p in packList:
    with open(os.path.join(packPath,p), "r") as read_file:
        data = json.load(read_file)
        if data["type"] != 'spell':
            continue
        if data["flags"]["a5e-for-dnd5e"].get("rareSpell"):
            data["folder"] = rareSpells[data["system"]["level"]]
        else:
            data["folder"] = spells[data["system"]["level"]]
    print(data["folder"])
    break

File: tools/spellFolders.mjs
import { promises as fs } from 'fs';
import path from 'path';
import { randomID } from './utils.mjs';
import yaml from 'js-yaml';
import ReferenceFolder from './ReferenceFolder.mjs';

const targetPack = 'spells';
const packPath = path.join('src', 'packs', targetPack);

const spellPack = await fs.readdir(packPath);

class Folder {
  /**
   *
   * @param {string} name
   * @param {object} context
   * @param {string | null} [context.parentFolder]
   * @param {string | null} [context.color]
   */
  constructor(name, { parentFolder, color, ...context } = {}) {
    this.name = name;
    this.color = color ?? null;
    this.flags = {};
    this.sort = 0;
    this.sorting = 'a';
    this._type = 'Item';
    this.folder = parentFolder ?? null;
    this._id = randomID();
    this._key = '!folders!' + this._id;
  }

  /**
   * Returns the folder as an object but with the _type replaced by type
   * @returns {object} The folder, as an object
   */
  get toObject() {
    const out = { ...this };
    out['type'] = this._type;
    delete out._type;
    return out;
  }
}

async function createFolders() {
  const outerFolders = ['Spells', 'Rare Spells'];
  const innerFolders = [
    {
      name: 'Cantrip',
      color: '#b18dcd',
    },
    {
      name: '1st Level',
      color: '#a27ebc',
    },
    {
      name: '2nd Level',
      color: '#926fac',
    },
    {
      name: '3rd Level',
      color: '#83609d',
    },
    {
      name: '4th Level',
      color: '#75528d',
    },
    {
      name: '5th Level',
      color: '#66447e',
    },
    {
      name: '6th Level',
      color: '#58366f',
    },
    {
      name: '7th Level',
      color: '#4a2960',
    },
    {
      name: '8th Level',
      color: '#3d1b52',
    },
    {
      name: '9th Level',
      color: '#300d44',
    },
  ];
  for (const o of outerFolders) {
    const outer = new Folder(o);
    let filename = outer.name.replace(' ', '_') + '_' + outer._id + '.yml';
    await fs.writeFile(
      path.join(packPath, filename),
      yaml.dump(outer.toObject, { indent: 2 }),
      { flag: 'w' }
    );
    for (const i of innerFolders) {
      const inner = new Folder(i.name, {
        parentFolder: outer._id,
        color: i.color,
      });
      filename =
        'folders_' + inner.name.replace(' ', '_') + '_' + inner._id + '.yml';
      await fs.writeFile(
        path.join(packPath, filename),
        yaml.dump(inner.toObject, { indent: 2 })
      );
    }
  }
  await ReferenceFolder.build(targetPack);
}
try {
  await fs.access(path.join('tools', 'referenceFolders', 'spells.yml'));
} catch {
  await createFolders();
}

const spells = Array.from('0123456789');
const rareSpells = Array.from('0123456789');

const folders = new ReferenceFolder(targetPack);

await folders.prepFolders();

for (const [id, f] of Object.entries(folders.compendium)) {
  if (
    !f.parent ||
    f.parent === 'XXlxppJh0OdnJaQQ' || // a5e-native "Spells" folder
    f.parent === 'ohddM2fcnkPeUeJv' // a5e-native "Rare spells" folder
  )
    continue;
  const parentName = folders.getFolderName(id, 1);
  const currentName = f.name;
  const level = currentName !== 'Cantrip' ? parseInt(currentName[0]) : 0;
  if (parentName === 'Spells') spells[level] = id;
  else rareSpells[level] = id;
}

for (const d of spellPack) {
  const read_file = await fs.readFile(path.join(packPath, d));
  const data = yaml.load(read_file);
  if (data.type !== 'spell') continue;
  if (data.flags['a5e-for-dnd5e']?.rareSpell) {
    data.folder = rareSpells[data.system.level];
  } else data.folder = spells[data.system.level];
  await fs.writeFile(
    path.join(packPath, d),
    yaml.dump(data, { indent: 2 }),
    'utf-8'
  );
}
